### PR TITLE
feat(matrix_references): finished matrix references and unit testing

### PR DESCRIPTION
**Pull request by mehdiwahada**

### CODE REVIEW

File: antarest/matrixstore/web.py [hunk 1]
```diff
@@ -97,6 +99,36 @@ def get(id: str) -> MatrixDTO:
             data=df.to_numpy().tolist(),
         )
 
+    @bp.get(
+        "/matrix/_references/{disk_usage}", tags=[APITag.matrix], description="Fetching a list of matrices statistics"
+    )
+    def get_matrices_references(disk_usage: bool) -> dict[str, MatrixReferencesDTO]:
+        logger.info("Fetching matrices references")
+        used_matrices = service.get_used_matrices()
```
Comment (TheoPascoli) [inline antarest/matrixstore/web.py]: All this business logic should be done inside the service. The "web" layer is just the entrypoint of the API, it gets the HTTP request, forward it to the service, then return the response. No business logic should be done here.
It is better because : 
- easier to test (no need to simulate http request to test this)
- reusable: sometimes we may might want to reuse some of the methods

As the others methods here, you should put most of the code here inside the service, and then call the service

File: antarest/matrixstore/web.py [hunk 2]
```diff
@@ -97,6 +99,36 @@ def get(id: str) -> MatrixDTO:
...
+
+        references_dto: dict[str, MatrixReferencesDTO] = {}
+
+        for matrix in used_matrices:
+            refs = []
+            matrix_size = None
+            matrix_id = matrix.matrix_id
+            if disk_usage:
+                matrix_size = service.get(matrix_id).__sizeof__()
```
Comment (TheoPascoli) [inline antarest/matrixstore/web.py]: I'm not really a fan of using `__` methods. As an alternative, you could write : 
```
import sys
matrix_size = sys.getsizeof(service.get(matrix_id))
```
Let me know what you think

File: antarest/matrixstore/web.py [hunk 3]
```diff
@@ -97,6 +99,36 @@ def get(id: str) -> MatrixDTO:
             data=df.to_numpy().tolist(),
         )
 
+    @bp.get(
+        "/matrix/_references/{disk_usage}", tags=[APITag.matrix], description="Fetching a list of matrices statistics"
```
Comment (sylvlecl) [inline antarest/matrixstore/web.py]: disk_usage should be a query parameter instead of a path parameter, you will need to annotate the function argument with `Query` (you can have a look at other endpoints for examples).

File: antarest/matrixstore/web.py [hunk 4]
```diff
@@ -97,6 +99,36 @@ def get(id: str) -> MatrixDTO:
...
+                matrix_size = service.get(matrix_id).__sizeof__()
+
+            description = matrix.use_description
+
+            ref_dto = MatrixDescriptionDTO.model_validate(MatrixDescriptionDTO(description=description))
+
+            refs.append(ref_dto)
+
+            if matrix_id not in references_dto:
+                refs_dto = MatrixReferencesDTO.model_validate({"refs": refs, "disk_usage": matrix_size})
```
Comment (sylvlecl) [inline antarest/matrixstore/web.py]: Here you may use the constructor instead of using a dict and model_validate:
`MatrixReferencesDTO(refs=refs, disk_usage=matrix_size)`

Comment (TheoPascoli) [inline antarest/matrixstore/web.py]: I don't think you should include disk_usage in the return object if the boolean is set to False. 
The response should not include None value.
For example, you can add a header to the FastAPI request like that : 
```
 @bp.get(
        "/matrix/_references/{disk_usage}", tags=[APITag.matrix], description="Fetching a list of matrices statistics", response_model_exclude_none=True
    )
```
Maybe there is also a possibility to do it at at the model level. sometimes we do something like model_validate(exclude_none=True).

File: antarest/matrixstore/web.py [hunk 5]
```diff
@@ -97,6 +99,36 @@ def get(id: str) -> MatrixDTO:
...
+        for matrix in used_matrices:
+            refs = []
+            matrix_size = None
+            matrix_id = matrix.matrix_id
+            if disk_usage:
+                matrix_size = service.get(matrix_id).__sizeof__()
+
+            description = matrix.use_description
+
+            ref_dto = MatrixDescriptionDTO.model_validate(MatrixDescriptionDTO(description=description))
```
Comment (sylvlecl) [inline antarest/matrixstore/web.py]: you can just use
```python
ref_dto = MatrixDescriptionDTO(description=description)
```

File: antarest/matrixstore/web.py [hunk 6]
```diff
@@ -97,6 +99,36 @@ def get(id: str) -> MatrixDTO:
...
+
+            refs.append(ref_dto)
```
Comment (sylvlecl) [inline antarest/matrixstore/web.py]: we'd better move it inside the `if` statement, it's not useful if the references object already exists

File: antarest/matrixstore/web.py [hunk 7]
```diff
@@ -97,6 +99,36 @@ def get(id: str) -> MatrixDTO:
...
+        "/matrix/_references/{disk_usage}", tags=[APITag.matrix], description="Fetching a list of matrices statistics"
+    )
+    def get_matrices_references(disk_usage: bool) -> dict[str, MatrixReferencesDTO]:
+        logger.info("Fetching matrices references")
+        used_matrices = service.get_used_matrices()
+
+        references_dto: dict[str, MatrixReferencesDTO] = {}
+
+        for matrix in used_matrices:
+            refs = []
```
Comment (sylvlecl) [inline antarest/matrixstore/web.py]: we'd better move it inside the `if` statement at line 124, it's not useful if the references object already exists.

Same for the lines regarding the disk_usage below.

**Review by TheoPascoli — CHANGES_REQUESTED** [review]
Hello Mehdi, good work overall, just a few things to change. Let me know if you have questions

File: antarest/matrixstore/web.py (hunk 2 above)
Comment (sylvlecl) [inline antarest/matrixstore/web.py]: actually here, what we want is really the size taken by the content of the matrix on disk, which is not what you get here.

To get that information, we could achieve it in 2 different ways:
- when the matrix is created, we get its size from disk and save it to the database. Then we can get it in the `MatrixMetadataDTO` from the matrix service. This is a good approach because we only compute the size once.
- a probably simpler approach is to add a function to the matrix content repository to get the disk usage of a matrix, since it knows where the matrices are stored on disk, and call that function here for each matrix

I propose that we start with the simpler approach.

Also, we'd better move that method to the matrix_service, because it's getting a little too complex for the API definition in web.py which should remain very basic.

**Review by sylvlecl — CHANGES_REQUESTED** [review]
Nice, a few things to change though:

- some suggestions about implementation details
- the disk_usage part needs to be changed to get the actual size of the file on disk

File: antarest/matrixstore/web.py [hunk 8]
```diff
@@ -97,6 +98,23 @@ def get(id: str) -> MatrixDTO:
...
+        tags=[APITag.matrix],
+        description="Fetching a list of matrices statistics",
+    )
+    def get_matrices_references(
+        disk_usage: bool = Query(
+            ...,
+            alias="disk_usage",
+            description="Determine if the disk usage should be displayed",
+            title="Disk Usage",
+            response_model_exclude_none=True,
```
Comment (TheoPascoli) [inline antarest/matrixstore/web.py]: I still have null value when setting `disk_usage` to false. You should move this line below tags and description

File: tests/matrixstore/test_web.py [hunk 9]
```diff
@@ -123,3 +123,86 @@ def test_import() -> None:
...
+        "constant_id_1": matrix_reference_two,
+        "test": matrix_reference_three,
+    }
+
+    service.get_used_matrices.return_value = matrix_references
+
+    service.get_matrices_references.return_value = matrix_dict
+
+    res = client.get("/v1/matrix/_references/", headers=create_auth_token(app), params={"disk_usage": False})
+    assert res.status_code == 200
```
Comment (TheoPascoli) [inline tests/matrixstore/test_web.py:192]: Can you also test the content of the http response ? To be sure that there is no disk_usage: null when it is set to false

File: antarest/matrixstore/service.py [hunk 10]
```diff
@@ -578,3 +593,30 @@ def get_matrix_usage(self) -> list[MatrixReference]:
...
+            matrix_size = None
+            matrix_id = matrix.matrix_id
+
+            description = matrix.use_description
+
+            ref_dto = MatrixDescriptionDTO(description=description)
+
+            if matrix_id not in references_dto:
+                refs = []
+                data = self.get(matrix_id)
```
Comment (TheoPascoli) [inline antarest/matrixstore/service.py]: The matrix_id **is** the matrix hash, you dont have to load the matrix to calcultate its hash, you can use the matrix_id

File: antarest/matrixstore/service.py [hunk 11]
```diff
@@ -578,3 +593,30 @@ def get_matrix_usage(self) -> list[MatrixReference]:
...
+
+    @override
+    def get_matrices_references(self, disk_usage: bool) -> dict[str, MatrixReferencesDTO]:
+        used_matrices = self.get_used_matrices()
+        references_dto: dict[str, MatrixReferencesDTO] = {}
+        for matrix in used_matrices:
+            matrix_size = None
+            matrix_id = matrix.matrix_id
+
+            description = matrix.use_description
```
Comment (TheoPascoli) [inline antarest/matrixstore/service.py]: You use this only one time, you can directly put `matrix.use_description` in the description

File: antarest/matrixstore/service.py [hunk 12]
```diff
@@ -578,3 +593,30 @@ def get_matrix_usage(self) -> list[MatrixReference]:
...
+
+            ref_dto = MatrixDescriptionDTO(description=description)
+
+            if matrix_id not in references_dto:
+                refs = []
```
Comment (TheoPascoli) [inline antarest/matrixstore/service.py]: You dont have to instanciate a list here because there can only be one element in it at the same time. 
1. You remove `refs = []`
2. You remove `refs.append(ref_dto)`
3. You write `refs_dto = MatrixReferencesDTO(refs=[ref_dto], disk_usage=matrix_size) `(that creates a list with one single element in it)

**Review by TheoPascoli — CHANGES_REQUESTED** [review]

File: antarest/matrixstore/repository.py [hunk 13]
```diff
@@ -336,3 +338,10 @@ def delete(self, matrix_hash: str) -> None:
         # Abandoned lock files are deleted here to maintain consistent behavior.
         lock_file = matrix_path.with_suffix(".tsv.lock")
         lock_file.unlink(missing_ok=True)
+
+    def get_matrix_disk_usage(self, matrix_hash: str) -> int:
+        for internal_format in InternalMatrixFormat:
```
Comment (sylvlecl) [inline antarest/matrixstore/repository.py:347]: we'd better extract a small method `_find_matrix_path` that will be used also by other functions that already perfom the same lookup

File: antarest/matrixstore/web.py [hunk 14]
```diff
@@ -97,6 +98,20 @@ def get(id: str) -> MatrixDTO:
             data=df.to_numpy().tolist(),
         )
 
+    @bp.get(
+        "/matrix/_references/",
+        tags=[APITag.matrix],
+        description="Fetching a list of matrices statistics",
+        response_model_exclude_none=True,
+    )
+    def get_matrices_references(
```
Comment (sylvlecl) [inline antarest/matrixstore/web.py:107]: Small stuff: this should be restricted to admin users

File: tests/matrixstore/test_web.py [hunk 15]
```diff
@@ -123,3 +123,103 @@ def test_import() -> None:
...
+                {"description": f"Referenced by command {command_id} of study study1"},
+                {"description": "Referenced by raw study study_id"},
+            ]
+        },
+        "constant_id_1": {"refs": [{"description": "Constant matrix"}]},
+        "test": {"refs": [{"description": "Referenced by dataset data_1"}]},
+    }
+    assert res.json() == data__
+
+    assert "disk_usage" not in res.json()[command_id]
```
Comment (sylvlecl) [inline tests/matrixstore/test_web.py]: You don't need those assertions since we already check the exact content of the response

File: tests/matrixstore/test_web.py [hunk 16]
```diff
@@ -123,3 +123,103 @@ def test_import() -> None:
...
+    assert "disk_usage" not in res.json()["constant_id_1"]
+    assert "disk_usage" not in res.json()["test"]
+
+    res_dict_test = creating_json_res_dict(res)
```
Comment (sylvlecl) [inline tests/matrixstore/test_web.py]: Same, I don't think we need all the following code since we have already checked the full result

File: tests/matrixstore/test_web.py [hunk 17]
```diff
@@ -123,3 +123,103 @@ def test_import() -> None:
...
+    app = create_app(service)
+    client = TestClient(app)
+    res = client.get("/v1/matrix/_references/", headers=create_auth_token(app), params={"disk_usage": True})
+
+    res_dict_test = creating_json_res_dict(res)
+
+    sorted_matrix_dict = dict(sorted(matrix_dict.items()))
+    sorted_res_matrix_dict = dict(sorted(res_dict_test.items()))
+    assert res.status_code == 200
+    assert sorted_res_matrix_dict == sorted_matrix_dict
```
Comment (sylvlecl) [inline tests/matrixstore/test_web.py:174]: I would be in favor of hard coding the expected result like below, it's less complicated to understand

**Review by sylvlecl — CHANGES_REQUESTED** [review]
Last minor stuff then it's OK !
In particular we need to restrict that feature to admin users, like the listing of all matrices (the GET previously implemented)

**Review by sylvlecl — APPROVED** [review]
